Block/Catalog/Product/View.php

```diff
@@ -90,84 +105,113 @@ public function __construct(
         WidgetConfigHelper $widgetConfigHelper,
         Config $config,
         Resolver $localeResolver,
+        Session $catalogSession,
+        Logger $logger,
+        ProductRepository $productRepository,
         array $data = []
     ) {
         parent::__construct($context, $data);
+        $this->logger = $logger;
+        $this->catalogSession = $catalogSession;
         $this->config = $config;
         $this->registry = $registry;
         $this->localeResolver = $localeResolver;
-        $this->getProduct();
-        $this->getPlans();
         $this->apiConfigHelper = $apiConfigHelper;
         $this->widgetConfigHelper = $widgetConfigHelper;
+        $this->productRepository = $productRepository;
+    }
+
+    /**
+     * Get last viewed product id session
+     * @return int
+     */
+    private function getLastViewProductId(): int
+    {
+        $this->logger->info('$this->catalogSession->getData()', [$this->catalogSession->getData()]);
+        return intval($this->catalogSession->getData('last_viewed_product_id'));
     }
 
     /**
-     * @return void
-     * @throws LocalizedException
+     * @return ProductInterface
+     * @throws AlmaProductException
      */
-    private function getProduct()
+    private function getProduct(): ProductInterface
     {
-        $this->product = $this->registry->registry('product');
-        if (!$this->product->getId()) {
-            throw new LocalizedException(__('Failed to initialize product'));
+        if (!is_null($this->product)) {
+            return $this->product;
         }
+
+        try {
+            $this->product = $this->productRepository->getById($this->getLastViewProductId());
+        } catch (NoSuchEntityException $e) {
+            $this->logger->error('No product in product Repository with this id', ['Exception message' => $e->getMessage(), 'Product ID' => $this->getLastViewProductId()]);
+            throw new AlmaProductException(sprintf('No product in product Repository with this id %s', $this->getLastViewProductId()));
+        }
+
+        if (is_null($this->product)) {
+            $this->product = $this->registry->registry('product');
```

Comment:
because `registry->registry('product')` is deprecated, could we test if `method_exists` before calling the method ?